net/tcp/tcp_timer.c

```diff
@@ -139,8 +140,10 @@ static void tcp_timer_expiry(FAR void *arg)
 {
   FAR struct tcp_conn_s *conn = arg;
 
+  net_lock();
```

Comment:
In general looks good, but I have a question. Is there a room for a race condition if `tcp_timer_expiry` is preempted by a thread running dev_poll so after returning back `arg` will no longer point to a valid `struct tcp_conn_s` structure? I mean maybe it is better to iterate
```
FAR struct tcp_conn_s *conn;

net_lock();

while ((conn = tcp_nextconn(conn)) != NULL && conn != arg);

if (conn != NULL)
  {
      conn->timeout = true;
      netdev_txnotify_dev(conn->dev);
  }

net_unlock();
```

Reply:
Done.

Reply:
Then maybe it is better to get rid of work queue per connection and have a single tcp_timer work queue that will run periodically with half second resolution and will integrate all the TCP connections and decrement timeout counters? What do you think?

Reply:
We are developing a wearable device need work for two weeks with 450mhA, so it isn't good to check the timeout every half second. That's why @zhhyu7 create https://github.com/apache/incubator-nuttx/pull/6355 in the first place.

Reply:
Performance value is zero if the system is multi threading unsafe. Let's merge current change and think about performance improvement. Maybe having a single work queue that is running at minimum of all connections timeout and iterating all the connections may be a solution instead of half a second periodic execution. I mean that it is a classic tick vs tickles mode problem, so we can apply a classic solution for it

Reply:
This is just another work queue, why not directly use work_s directly like #6355? The additional cost is very small since network stack already depend on the system work.

Reply:
Let me analyze a bit more

Reply:
> This is just another work queue, why not directly use work_s directly like #6355? The additional cost is very small since network stack already depend on the system work.

Could you please expand the description. I'm not fully tracking what you are trying to say.
I see that `struct tcp_conn_s` was equipped with `struct work_s work; /* TCP timer handle */` in https://github.com/apache/incubator-nuttx/pull/6330 so we start getting work queue per TCP connecting instead of a single work queue polling `devif_timer`. Then a polling work queue has been removed by https://github.com/apache/incubator-nuttx/pull/6355. Is that correct?

If above is "yes", then why we can't remove `struct work_s` from `struct tcp_conn_s` and adding a global `struct work_s tcp_timer_work;`. Each `truct tcp_conn_s` will have `int time_left;` field. The `tcp_timer_work` will `tcp_nextconn(conn)` and do `conn->time_left -= delay;` and `conn->timeout = true;` if delay is `0`. The `tcp_timer_work` can be started passing the delay via `arg`, so each `tcp_update_timer` call can check and adjust `tcp_timer_work` is connection timeout is less than time left since work queue is scheduled.

Maybe I'm missing some points, so please help me to understand.

Reply:
> > This is just another work queue, why not directly use work_s directly like #6355? The additional cost is very small since network stack already depend on the system work.
> 
> Could you please expand the description. I'm not fully tracking what you are trying to say. I see that `struct tcp_conn_s` was equipped with `struct work_s work; /* TCP timer handle */` in #6330 so we start getting work queue per TCP connecting instead of a single work queue polling `devif_timer`. Then a polling work queue has been removed by #6355. Is that correct?
> 

Yes, but the term usage isn't accurate. Both still use the same(LPWORK) work queue. The difference is that the old use one work_s for each netdev driver, but the new use one work_s for each tcp conn. 
Note: work_s isn't work queue, which is a simple struct(contain timeout, callback etc), not 1:1 mapping to thread.

> If above is "yes", then why we can't remove `struct work_s` from `struct tcp_conn_s` and adding a global `struct work_s tcp_timer_work;`. Each `truct tcp_conn_s` will have `int time_left;` field. The `tcp_timer_work` will `tcp_nextconn(conn)` and do `conn->time_left -= delay;` and `conn->timeout = true;` if delay is `0`. The `tcp_timer_work` can be started passing the delay via `arg`, so each `tcp_update_timer` call can check and adjust `tcp_timer_work` is connection timeout is less than time left since work queue is scheduled.
> 

This is just how the work queue work:
https://github.com/apache/incubator-nuttx/blob/master/sched/wqueue/kwork_thread.c#L156-L180

> Maybe I'm missing some points, so please help me to understand.

My point is that we can reuse sched/wqueue code, there isn't any benefit to write a dedicated work queue for network stack.

Reply:
No. I'm not talking about separate sched/wqueue for network stack, but reusing an LPWORK for that, but adding a single TCP work function to LPWORK queue that will iterate all the connections with lock taken. I do not see how we have achieve multi thread safety and good performance in other way.
```
          leave_critical_section(flags);
          CALL_WORKER(worker, arg);
          flags = enter_critical_section();
```
The preemption may happen right after `leave_critical_section(flags);` and before `CALL_WORKER(worker, arg);`, so do not see how we can rely on `arg` in `tcp_timer_expiry` without a `tcp_nextconn()` loop. But that seems to give performance penalty, so we need to conduct a more lightweight solution.

Reply:
Ok. So let's keep the current version.